### ISSUE TITLE
Add support of `on conflict` to the session.add method

### ISSUE THREAD

**Issue by suvorovalexander** [sqlalchemy/sqlalchemy#4448]
Currently there is no way to specify `on conflict` behaviour when using `session` class and it's `add` method which makes a decision to either `insert` or `update`. 
This would be a nice addition to have `on conflict` behaviour of this method configurable.

**Comment by zzzeek** [comment]
also can you read the docuemtnation for Session.merge() and tell me how *functionally* this differs from what I think you intend "conflict" to do?   if I wasn't watching SQL go by, how would your new feature differ? https://docs.sqlalchemy.org/en/latest/orm/session_api.html?highlight=session%20merge#sqlalchemy.orm.session.Session.merge

**Comment by suvorovalexander** [comment]
If I understand it correctly merge is only able to cover the case of when the primary key is violated.
And if I'm not mistaking it would also could be affect by the race condition.

Postgres supports `insert ... on conflict` statements.  where `on conflict` part is executed whenever any on constraints are violated. 

so the signature could be `session.add(<entity_obj>, on_conflict=<sql_statement>)`
e.g `session.add(user, on_conflict=do_nothing)` or `session.add(user, on_conflict=update({'email': 42)`.

**Comment by zzzeek** [comment]
Unfortunately this would be massively complicated not just to implement, but would not be possible to define from a behavioral standpoint without throwing away the whole ORM.  an ORM object is a lot more than just "a row" because it also refers to related items as well as lots of lifecycle events, and is also part of an identity map where the identity of an object is a maintained as unique.

Suppose I have two objects:

   m1 = Object(some_condition=A)
   m2 = Object(some_condition=B)

then suppose the ON CONFLICT I'm using means that m1 and m2 end up being the same row:

    session.add(m1, on_conflict="A")
    session.add(m2, on_conflict="A")

now flush.   Which of m1 or m2 goes into the identity map?  Which one gets thrown away?  how does the user know which one gets thrown away?      What if m1 and m2 both refer to "m1.related_item", where "m1.related_item" and "m2.related_item" are different?  which one wins?         It goes against the contract of session.add() that it can be called multiple times for the same identity which is what this case calls for.    Rest assured there would be many more issues of this nature that don't work with the ORM's usage contract.

The most that an ON CONFLICT style of SQL statement could be integrated into the unit of work and identity map would be that session.merge() could make use of a primary-key based conflict statement if the underlying database supports it. 

The above issues also don't accommodate for the fact that "ON CONFLICT" is a Postgresql-specific construct that is not available on any other database in the same form.    MySQL has "ON DUPLICATE KEY UPDATE" which acts totally differently, and then for other databases such as SQLIte, SQL Server, Oracle, they support still other forms, in the case of Oracle they only support MERGE which SQLAlchemy has no support for.   The bigger issue of implementing a database-agnostic style of merge is issue #960.

Finally, if an approach to all the above issues could be decided upon, there would need to be developers willing to take on these tasks, and as evidenced in #960, nobody is interested in stepping up even for the vastly simpler task of implementing a cross-platform merge construct in Core.

Unfortunately this request is much much more involved than it might appear at first.    If you are interested in trying to work up a specification that tries to deal with all of the above issues to define a workable usage contract I can leave this open but I have no plans to work on this.

**Comment by zzzeek** [comment]
note that key limitation with non-pk based "on conflict" is that the ORM needs the primary key for rows at all times.  MySQL has ON DUPLICATE KEY UPDATE but has no RETURNING so we would have no way to locate the row.    This is why this use case is so specific to Postgresql.     There is also the issue of how to merge all the related items, if not on primary key - this would need to be part of the API.    Also the add() issue above I refer to is solved by merge() because it does the SQL immediately you get a new object back.  

So to sum up:

1. ON CONFLICT type functionality in session.merge(), or a similar function - possible, but lots of work and tests, it would not be available on all backends.   if conflict resolution is configurable then it need a system of noting this for all related paths as well.

2. ON CONFLICT type functionality in add(), takes place during flush() - totally changes the whole ORM paradigm, vastly more difficult to deal with all the edge cases.

**Comment by charleswilmot** [comment]
If anyone needs a workaround, here is what I went for personally:

```
from sqlalchemy.dialects.sqlite import insert

with Session(engine) as session:
    ...
    statement = insert(SomeTable).values([dataclasses.asdict(data)]).on_conflict_do_nothing()
    session.execute(statement)
```

A simple  `session.add(data, on_conflict=db.do_nothing)` would be nicer indeed, but keeping the code conceptually logical is more important

**Comment by mw3i** [comment]
worth nothing that Django's ORM function `bulk_create(...)` has an `update_conflicts = ...` parameter that's incredibly convenient

**Comment by zzzeek** [comment]
> worth nothing that Django's ORM function `bulk_create(...)` has an `update_conflicts = ...` parameter that's incredibly convenient

what does it do on the backend?  I think our feature at https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-upsert-statements is very appropriate for us, users want to know that a real UPSERT is being used and not some kind of "load every row" thing that can't be easily caught.

**Comment by mw3i** [comment]
> 

good point! I don't know what it's doing on the backend, but your concern that it might be "faking" an UPSERT is valid

**Comment by nitedani** [comment]
> If anyone needs a workaround, here is what I went for personally:
> 
> ```
> from sqlalchemy.dialects.sqlite import insert
> 
> with Session(engine) as session:
>     ...
>     statement = insert(SomeTable).values([dataclasses.asdict(data)]).on_conflict_do_nothing()
>     session.execute(statement)
> ```
> 
> A simple `session.add(data, on_conflict=db.do_nothing)` would be nicer indeed, but keeping the code conceptually logical is more important

This doesn't work for me. 
```
null value in column "id" of relation "some_table" violates not-null constraint
```

**Comment by CaselIT** [comment]
you seem to be passing a None in the id. If that is a primary key it will not work
